xarray/tests/test_dataset.py

```diff
@@ -4036,6 +4036,40 @@ def data_set(seed=None):
     return create_test_data(seed)
 
 
+@pytest.mark.parametrize('unaligned_coords', (
+    {'x': [2, 1, 0]},
+    {'x': (['x'], np.asarray([2, 1, 0]))},
+    {'x': (['x'], np.asarray([1, 2, 0]))},
+    {'x': pd.Index([2, 1, 0])},
+    {'x': Variable(dims='x', data=[0, 2, 1])},
+    {'y': 42},
+    {'y': ('x', [2, 1, 0])},
+    {'y': ('x', np.asarray([2, 1, 0]))},
+    {'y': (['x'], np.asarray([2, 1, 0]))},
+))
+def test_constructor_aligns_to_explicit_coords(unaligned_coords):
+
+    coords = {'x': [0, 1, 2]}
+    a = xr.DataArray([1, 2, 3], dims=['x'], coords=unaligned_coords)
+
+    expected = xr.Dataset(coords=coords)
+    expected['a'] = a
+
+    result = xr.Dataset({'a': a}, coords=coords)
```

Comment:
Here's a case that breaks with your patch:
```
In [22]: array = xr.DataArray([1, 2, 3], [('x', [0, 1, 2])])

In [23]: xr.Dataset({'foo': array}, {'x': ('x', array.x.values)})
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-23-6c1c4b1b1cab> in <module>()
----> 1 xr.Dataset({'foo': array}, {'x': ('x', array.x.values)})

~/dev/xarray/xarray/core/dataset.py in __init__(self, data_vars, coords, attrs, compat)
    362             coords = {}
    363         if data_vars is not None or coords is not None:
--> 364             self._set_init_vars_and_dims(data_vars, coords, compat)
    365         if attrs is not None:
    366             self.attrs = attrs

~/dev/xarray/xarray/core/dataset.py in _set_init_vars_and_dims(self, data_vars, coords, compat)
    380
    381         variables, coord_names, dims = merge_data_and_coords(
--> 382             data_vars, coords, compat=compat)
    383
    384         self._variables = variables

~/dev/xarray/xarray/core/merge.py in merge_data_and_coords(data, coords, compat, join)
    361     objs = [data, coords]
    362     explicit_coords = coords.keys()
--> 363     return merge_core(objs, compat, join, explicit_coords=explicit_coords, indexes=coords)
    364
    365

~/dev/xarray/xarray/core/merge.py in merge_core(objs, compat, join, priority_arg, explicit_coords, indexes)
    422
    423     coerced = coerce_pandas_values(objs)
--> 424     aligned = deep_align(coerced, join=join, copy=False, indexes=indexes)
    425     expanded = expand_variable_dicts(aligned)
    426

~/dev/xarray/xarray/core/alignment.py in deep_align(objects, join, copy, indexes, exclude, raise_on_invalid)
    211
    212     aligned = align(*targets, join=join, copy=copy, indexes=indexes,
--> 213                     exclude=exclude)
    214
    215     for position, key, aligned_obj in zip(positions, keys, aligned):

~/dev/xarray/xarray/core/alignment.py in align(*objects, **kwargs)
    118     for dim, matching_indexes in iteritems(all_indexes):
    119         if dim in indexes:
--> 120             index = utils.safe_cast_to_index(indexes[dim])
    121             if (any(not index.equals(other) for other in matching_indexes) or
    122                     dim in unlabeled_dim_sizes):

~/dev/xarray/xarray/core/utils.py in safe_cast_to_index(array)
     54         if hasattr(array, 'dtype') and array.dtype.kind == 'O':
     55             kwargs['dtype'] = object
---> 56         index = pd.Index(np.asarray(array), **kwargs)
     57     return index
     58

~/miniconda3/lib/python3.6/site-packages/numpy/core/numeric.py in asarray(a, dtype, order)
    490
    491     """
--> 492     return array(a, dtype, copy=False, order=order)
    493
    494

ValueError: setting an array element with a sequence
```

Your test case didn't catch this because it was only changing `coords` on the DataArray argument, not `coords` for the Dataset.